source/AI.cpp

```diff
@@ -308,7 +349,126 @@ AI::AI(const List<Ship> &ships, const List<Minable> &minables, const List<Flotsa
 
 
 
+
+
+
+
+// Fleet commands from the player.
+void AI::IssueFormationChange(const PlayerInfo &player)
+{
+	// Figure out what ships we are giving orders to
+	vector<Ship *> targetShips = GetShipsForFormationCommand(player);
+	
+	// This should never happen, but just in case:
+	if(targetShips.empty())
+		return;
+	
+	const auto &formationPatterns = GameData::Formations();
+	if(formationPatterns.size() < 1)
+	{
+		Messages::Add("No formations available.", Messages::Importance::High);
+		return;
+	}
+	
+	// First check which and how many formation patterns we have in the current set of selected ships.
+	// If there is more than 1 pattern in the selection, then this command will change all selected
+	// ships to use the pattern that was found first. If there is only 1 pattern, then this command
+	// will change all selected ships to use the next pattern available.
+	const FormationPattern *toSet = nullptr;
+	bool multiplePatternsSet = false;
+	for(Ship *ship : targetShips)
+	{
+		const FormationPattern *shipsPattern = ship->GetFormationPattern();
+		if(shipsPattern)
+		{
+			if(!toSet)
+				toSet = shipsPattern;
+			else if(toSet != shipsPattern)
+			{
+				multiplePatternsSet = true;
+				break;
+			}
+		}
+	}
+	
+	// Now determine what formationPattern to set.
+	if(!toSet)
+		// If no pattern was set at all, then we set the first one from the set of formationPatterns.
+		toSet = &(formationPatterns.begin()->second);
+	else if(!multiplePatternsSet)
+	{
+		// If only one pattern was found, then select the next pattern (or clear the pattern if there is no next).
+		auto it = formationPatterns.find(toSet->Name());
+		if(it != formationPatterns.end())
+			++it;
+		toSet = (it == formationPatterns.end() ? nullptr : &(it->second));
+	}
+	//else if (multiplePatternsSet) {
+	// If more than one formation was found on the ships, then we set the pattern it to the first one found.
+	// No code is needed here for this option, since all variables are already set to just apply the change below.
+	//}
+	
+	// Now set the pattern on the selected ships.
+	for(Ship *ship : targetShips)
+		if(ship->GetFormationPattern() != toSet)
+			ship->SetFormationPattern(toSet);
+	
+	if(toSet)
+		Messages::Add(to_string(targetShips.size()) + " ships are now flying in " + toSet->Name() + " formation.", Messages::Importance::Low);
+	else
+		Messages::Add(to_string(targetShips.size()) + " ships are no longer flying in formation.", Messages::Importance::Low);
+}
+
+
+
+// Fleet commands from the player.
+void AI::IssueFormationRingDecrease(const PlayerInfo &player)
```

Comment:
Good idea. Those controls were based on some initial controls used for debugging and zeroing was more useful back then. But it looks like those controls are going to be official, and in that case we want the decrease function to just perform decrease.
I will also combine some code from the Decrease and Increase functions (since they will be almost identical).